File: windows/src/Core/FountainCodec.cs
using System;
using System.Collections.Generic;

namespace OpticalTransfer.Core
{
    public enum DecoderResult
    {
        NeedMoreSymbols,
        SymbolIngested,
        DuplicateSymbol,
        Solved,
        CorruptedSymbol
    }

    public class DecoderStatus
    {
        public uint K { get; set; }
        public uint SolvedCount { get; set; }
        public uint ReceivedCount { get; set; }
        public float ProgressPercentage { get; set; }
        public bool IsComplete { get; set; }
        public uint SessionId { get; set; }
    }

    public class SystematicFountainEncoder
    {
        private readonly byte[] _payload;
        private readonly ushort _symbolSize;
        private readonly uint _k;
        private readonly uint _sessionId;
        private readonly double[] _cdf;
        private uint _currentSeq;

        public SystematicFountainEncoder(byte[] data, ushort targetSymbolSize = 300)
        {
            _payload = data ?? new byte[0];
            _symbolSize = targetSymbolSize;
            _currentSeq = 0;

            if (_payload.Length == 0)
            {
                _k = 1;
            }
            else
            {
                _k = (uint)((_payload.Length + _symbolSize - 1) / _symbolSize);
                if (_k == 0) _k = 1;
            }

            Random rnd = new Random();
            _sessionId = (uint)rnd.Next(1, 65535);
            _cdf = FountainMath.SolitonCdf(_k);
        }

        public uint K { get { return _k; } }
        public ushort SymbolSize { get { return _symbolSize; } }
        public uint SessionId { get { return _sessionId; } }
        public uint TotalBytes { get { return (uint)_payload.Length; } }
        public uint CurrentSeq { get { return _currentSeq; } }

        public byte[] NextFrame()
        {
            uint seq = _currentSeq++;
            byte[] blockData = new byte[_symbolSize];

            if (seq < _k)
            {
                // Systematic transmission: direct chunk of source payload
                int start = (int)seq * _symbolSize;
                int end = Math.Min(start + _symbolSize, _payload.Length);
                if (start < _payload.Length)
                {
                    Buffer.BlockCopy(_payload, start, blockData, 0, end - start);
                }
            }
            else
            {
                // Repair transmission: Soliton-selected linear combination (XOR)
                uint[] indices = FountainMath.FrameIndices(_k, _cdf, _sessionId, seq);
                foreach (uint idx in indices)
                {
                    int start = (int)idx * _symbolSize;
                    int end = Math.Min(start + _symbolSize, _payload.Length);
                    if (start < _payload.Length)
                    {
                        byte[] temp = new byte[_symbolSize];
                        Buffer.BlockCopy(_payload, start, temp, 0, end - start);
                        FountainMath.XorBlocks(blockData, temp, end - start);
                    }
                }
            }

            OtiFrameHeader header = new OtiFrameHeader
            {
                Magic = ProtocolDcf2.OTI_MAGIC,
                SessionId = (ushort)(_sessionId & 0xFFFF),
                TotalPayloadBytes = (uint)_payload.Length,
                SymbolSizeBytes = _symbolSize,
                TotalSourceSymbols = (ushort)(_k & 0xFFFF),
                Seq = (ushort)(seq & 0xFFFF),
                HeaderCrc16 = 0
            };

            return ProtocolDcf2.PackFrame(header, blockData);
        }
    }

    public class IncrementalPeelingDecoder
    {
        private class PendingEquation
        {
            public HashSet<uint> Indices = new HashSet<uint>();
            public byte[] Data;
        }

        private uint _sessionId;
        private uint _k;
        private ushort _symbolSize;
        private uint _totalPayloadBytes;
        private double[] _cdf;

        private readonly Dictionary<uint, byte[]> _solvedBlocks = new Dictionary<uint, byte[]>();
        private readonly Dictionary<uint, List<PendingEquation>> _byBlockMap = new Dictionary<uint, List<PendingEquation>>();
        private readonly HashSet<uint> _seenSeqs = new HashSet<uint>();
        private uint _receivedFramesCount;

        public IncrementalPeelingDecoder()
        {
            Reset();
        }

        public void Reset()
        {
            _sessionId = 0;
            _k = 0;
            _symbolSize = 0;
            _totalPayloadBytes = 0;
            _cdf = null;
            _solvedBlocks.Clear();
            _byBlockMap.Clear();
            _seenSeqs.Clear();
            _receivedFramesCount = 0;
        }

        public bool IsComplete
        {
            get { return _k > 0 && _solvedBlocks.Count >= _k; }
        }

        public DecoderResult ConsumeFrame(byte[] frameBytes)
        {
            OtiFrameHeader header;
            byte[] blockData;
            if (!ProtocolDcf2.ParseFrame(frameBytes, out header, out blockData))
            {
                return DecoderResult.CorruptedSymbol;
            }

            // Reset if session salt changes
            if (_sessionId != 0 && (header.SessionId != _sessionId || header.TotalSourceSymbols != _k || header.TotalPayloadBytes != _totalPayloadBytes))
            {
                Reset();
            }

            if (_sessionId == 0)
            {
                _sessionId = header.SessionId;
                _k = header.TotalSourceSymbols;
                _symbolSize = header.SymbolSizeBytes;
                _totalPayloadBytes = header.TotalPayloadBytes;
                _cdf = FountainMath.SolitonCdf(_k);
            }

            if (_seenSeqs.Contains(header.Seq))
            {
                return DecoderResult.DuplicateSymbol;
            }
            _seenSeqs.Add(header.Seq);
            _receivedFramesCount++;

            if (IsComplete)
            {
                return DecoderResult.Solved;
            }

            HashSet<uint> indices = new HashSet<uint>();
            if (header.Seq < _k)
            {
                indices.Add(header.Seq);
            }
            else
            {
                uint[] indVec = FountainMath.FrameIndices(_k, _cdf, _sessionId, header.Seq);
                foreach (uint u in indVec) indices.Add(u);
            }

            byte[] currentPayload = new byte[_symbolSize];
            Buffer.BlockCopy(blockData, 0, currentPayload, 0, _symbolSize);

            // Eliminate already solved blocks
            List<uint> toRemove = new List<uint>();
            foreach (uint b in indices)
            {
                if (_solvedBlocks.ContainsKey(b))
                {
                    FountainMath.XorBlocks(currentPayload, _solvedBlocks[b], _symbolSize);
                    toRemove.Add(b);
                }
            }
            foreach (uint b in toRemove)
            {
                indices.Remove(b);
            }

            if (indices.Count == 0)
            {
                return IsComplete ? DecoderResult.Solved : DecoderResult.SymbolIngested;
            }

            if (indices.Count == 1)
            {
                uint solvedIdx = 0;
                foreach (uint idx in indices) { solvedIdx = idx; break; }
                ResolveBlock(solvedIdx, currentPayload);
            }
            else
            {
                PendingEquation eq = new PendingEquation
                {
                    Indices = indices,
                    Data = currentPayload
                };

                foreach (uint b in eq.Indices)
                {
                    if (!_byBlockMap.ContainsKey(b))
                    {
                        _byBlockMap[b] = new List<PendingEquation>();
                    }
                    _byBlockMap[b].Add(eq);
                }
            }

            return IsComplete ? DecoderResult.Solved : DecoderResult.SymbolIngested;
        }

        private void ResolveBlock(uint b0, byte[] w0)
        {
            Queue<KeyValuePair<uint, byte[]>> queue = new Queue<KeyValuePair<uint, byte[]>>();
            queue.Enqueue(new KeyValuePair<uint, byte[]>(b0, w0));

            while (queue.Count > 0)
            {
                KeyValuePair<uint, byte[]> item = queue.Dequeue();
                uint b = item.Key;
                byte[] w = item.Value;

                if (_solvedBlocks.ContainsKey(b)) continue;
                _solvedBlocks[b] = w;

                if (!_byBlockMap.ContainsKey(b)) continue;

                List<PendingEquation> waitingList = _byBlockMap[b];
                _byBlockMap.Remove(b);

                foreach (PendingEquation eq in waitingList)
                {
                    if (eq.Indices.Contains(b))
                    {
                        FountainMath.XorBlocks(eq.Data, w, _symbolSize);
                        eq.Indices.Remove(b);

                        if (eq.Indices.Count == 1)
                        {
                            uint resolvedIdx = 0;
                            foreach (uint idx in eq.Indices) { resolvedIdx = idx; break; }
                            if (!_solvedBlocks.ContainsKey(resolvedIdx))
                            {
                                byte[] copyData = new byte[_symbolSize];
                                Buffer.BlockCopy(eq.Data, 0, copyData, 0, _symbolSize);
                                queue.Enqueue(new KeyValuePair<uint, byte[]>(resolvedIdx, copyData));
                            }
                        }
                    }
                }
            }
        }

        public DecoderStatus GetStatus()
        {
            return new DecoderStatus
            {
                K = _k,
                SolvedCount = (uint)_solvedBlocks.Count,
                ReceivedCount = _receivedFramesCount,
                ProgressPercentage = (_k > 0) ? ((float)_solvedBlocks.Count / (float)_k * 100.0f) : 0.0f,
                IsComplete = IsComplete,
                SessionId = _sessionId
            };
        }

        public byte[] AssemblePayload()
        {
            if (!IsComplete) return null;

            byte[] payload = new byte[_totalPayloadBytes];
            for (uint b = 0; b < _k; ++b)
            {
                if (!_solvedBlocks.ContainsKey(b)) return null;

                int start = (int)b * _symbolSize;
                int len = Math.Min((int)_symbolSize, (int)_totalPayloadBytes - start);
                if (len > 0)
                {
                    Buffer.BlockCopy(_solvedBlocks[b], 0, payload, start, len);
                }
            }
            return payload;
        }
    }
}

File: native/core/src/fountain_codec.cpp
#include "fountain_codec.hpp"
#include <random>
#include <cstring>
#include <algorithm>

namespace Optical::Core {

SystematicFountainEncoder::SystematicFountainEncoder(const uint8_t* data, size_t len, uint16_t target_symbol_size)
    : payload_(data, data + len),
      symbol_size_(target_symbol_size),
      current_seq_(0)
{
    if (payload_.empty()) {
        k_ = 1;
    } else {
        k_ = static_cast<uint32_t>((payload_.size() + symbol_size_ - 1) / symbol_size_);
        if (k_ == 0) k_ = 1;
    }

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<uint32_t> dis(1, 65535);
    session_id_ = dis(gen);

    cdf_ = FountainMath::solitonCdf(k_);
}

std::vector<uint8_t> SystematicFountainEncoder::nextFrame() {
    uint32_t seq = current_seq_++;

    std::vector<uint8_t> block_data(symbol_size_, 0);

    if (seq < k_) {
        // Systematic transmission: direct chunk of source payload
        size_t start = static_cast<size_t>(seq) * symbol_size_;
        size_t end = std::min(start + symbol_size_, payload_.size());
        if (start < payload_.size()) {
            std::memcpy(block_data.data(), payload_.data() + start, end - start);
        }
    } else {
        // Repair transmission: Soliton-selected linear combination (XOR)
        std::vector<uint32_t> indices = FountainMath::frameIndices(k_, cdf_, session_id_, seq);
        for (uint32_t idx : indices) {
            size_t start = static_cast<size_t>(idx) * symbol_size_;
            size_t end = std::min(start + symbol_size_, payload_.size());
            if (start < payload_.size()) {
                FountainMath::xorBlocks(block_data.data(), payload_.data() + start, end - start);
            }
        }
    }

    OtiFrameHeader header;
    header.magic = OTI_MAGIC;
    header.session_id = static_cast<uint16_t>(session_id_ & 0xFFFF);
    header.total_payload_bytes = static_cast<uint32_t>(payload_.size());
    header.symbol_size_bytes = symbol_size_;
    header.total_source_symbols = static_cast<uint16_t>(k_ & 0xFFFF);
    header.seq = static_cast<uint16_t>(seq & 0xFFFF);
    header.header_crc16 = 0;

    return Protocol::packFrame(header, block_data.data());
}

IncrementalPeelingDecoder::IncrementalPeelingDecoder() {
    reset();
}

void IncrementalPeelingDecoder::reset() {
    session_id_ = 0;
    k_ = 0;
    symbol_size_ = 0;
    total_payload_bytes_ = 0;
    cdf_.clear();
    solved_blocks_.clear();
    by_block_map_.clear();
    all_equations_.clear();
    seen_seqs_.clear();
    received_frames_count_ = 0;
}

DecoderResult IncrementalPeelingDecoder::consumeFrame(const uint8_t* frame_bytes, size_t len) {
    OtiFrameHeader header;
    const uint8_t* block_data = nullptr;
    if (!Protocol::parseFrame(frame_bytes, len, header, block_data)) {
        return DecoderResult::CorruptedSymbol;
    }

    // Reset if session salt changes
    if (session_id_ != 0 && (header.session_id != session_id_ || header.total_source_symbols != k_ || header.total_payload_bytes != total_payload_bytes_)) {
        reset();
    }

    if (session_id_ == 0) {
        session_id_ = header.session_id;
        k_ = header.total_source_symbols;
        symbol_size_ = header.symbol_size_bytes;
        total_payload_bytes_ = header.total_payload_bytes;
        cdf_ = FountainMath::solitonCdf(k_);
    }

    if (seen_seqs_.count(header.seq) > 0) {
        return DecoderResult::DuplicateSymbol;
    }
    seen_seqs_.insert(header.seq);
    received_frames_count_++;

    if (isComplete()) {
        return DecoderResult::Solved;
    }

    std::set<uint32_t> indices;
    if (header.seq < k_) {
        indices.insert(header.seq);
    } else {
        std::vector<uint32_t> ind_vec = FountainMath::frameIndices(k_, cdf_, session_id_, header.seq);
        indices.insert(ind_vec.begin(), ind_vec.end());
    }

    std::vector<uint8_t> current_payload(block_data, block_data + symbol_size_);

    // Eliminate already solved blocks
    auto it = indices.begin();
    while (it != indices.end()) {
        uint32_t b = *it;
        auto solved_it = solved_blocks_.find(b);
        if (solved_it != solved_blocks_.end()) {
            FountainMath::xorBlocks(current_payload.data(), solved_it->second.data(), symbol_size_);
            it = indices.erase(it);
        } else {
            ++it;
        }
    }

    if (indices.empty()) {
        return isComplete() ? DecoderResult::Solved : DecoderResult::SymbolIngested;
    }

    if (indices.size() == 1) {
        resolveBlock(*indices.begin(), current_payload);
    } else {
        auto eq = std::make_shared<PendingEquation>();
        eq->indices = std::move(indices);
        eq->data = std::move(current_payload);
        all_equations_.push_back(eq);

        for (uint32_t b : eq->indices) {
            by_block_map_[b].push_back(eq);
        }

        // Trigger sparse Gaussian elimination if we have enough symbols but peeling stalled
        if (received_frames_count_ >= k_ + 2 && !isComplete()) {
            runGaussianEliminationFallback();
        }
    }

    return isComplete() ? DecoderResult::Solved : DecoderResult::SymbolIngested;
}

void IncrementalPeelingDecoder::resolveBlock(uint32_t b0, const std::vector<uint8_t>& w0) {
    std::deque<std::pair<uint32_t, std::vector<uint8_t>>> queue;
    queue.push_back({b0, w0});

    while (!queue.empty()) {
        auto [b, w] = queue.front();
        queue.pop_front();

        if (solved_blocks_.count(b) > 0) continue;
        solved_blocks_[b] = w;

        auto waiting_it = by_block_map_.find(b);
        if (waiting_it == by_block_map_.end()) continue;

        auto waiting_list = std::move(waiting_it->second);
        by_block_map_.erase(waiting_it);

        for (auto& eq : waiting_list) {
            FountainMath::xorBlocks(eq->data.data(), w.data(), symbol_size_);
            eq->indices.erase(b);

            if (eq->indices.size() == 1) {
                uint32_t resolved_idx = *eq->indices.begin();
                if (solved_blocks_.count(resolved_idx) == 0) {
                    queue.push_back({resolved_idx, eq->data});
                }
            }
        }
    }
}

bool IncrementalPeelingDecoder::runGaussianEliminationFallback() {
    // Filter active unresolved equations
    std::vector<std::shared_ptr<PendingEquation>> active_eqs;
    for (auto& eq : all_equations_) {
        if (!eq->indices.empty()) {
            active_eqs.push_back(eq);
        }
    }

    if (active_eqs.empty()) return false;

    // Fast sparse elimination pass
    for (size_t i = 0; i < active_eqs.size(); ++i) {
        if (active_eqs[i]->indices.empty()) continue;
        uint32_t pivot = *active_eqs[i]->indices.begin();

        for (size_t j = i + 1; j < active_eqs.size(); ++j) {
            if (active_eqs[j]->indices.count(pivot) > 0) {
                // XOR row j with row i
                for (uint32_t idx : active_eqs[i]->indices) {
                    if (active_eqs[j]->indices.count(idx) > 0) {
                        active_eqs[j]->indices.erase(idx);
                    } else {
                        active_eqs[j]->indices.insert(idx);
                    }
                }
                FountainMath::xorBlocks(active_eqs[j]->data.data(), active_eqs[i]->data.data(), symbol_size_);

                if (active_eqs[j]->indices.size() == 1) {
                    resolveBlock(*active_eqs[j]->indices.begin(), active_eqs[j]->data);
                }
            }
        }
    }
    return isComplete();
}

bool IncrementalPeelingDecoder::isComplete() const {
    return k_ > 0 && solved_blocks_.size() >= k_;
}

DecoderStatus IncrementalPeelingDecoder::getStatus() const {
    DecoderStatus status;
    status.k = k_;
    status.solved_count = static_cast<uint32_t>(solved_blocks_.size());
    status.received_count = received_frames_count_;
    status.progress_percentage = (k_ > 0) ? (static_cast<float>(solved_blocks_.size()) / static_cast<float>(k_) * 100.0f) : 0.0f;
    status.is_complete = isComplete();
    status.session_id = session_id_;
    return status;
}

std::optional<std::vector<uint8_t>> IncrementalPeelingDecoder::assemblePayload() const {
    if (!isComplete()) return std::nullopt;

    std::vector<uint8_t> payload(total_payload_bytes_);
    for (uint32_t b = 0; b < k_; ++b) {
        auto it = solved_blocks_.find(b);
        if (it == solved_blocks_.end()) return std::nullopt;

        size_t start = static_cast<size_t>(b) * symbol_size_;
        size_t len = std::min(static_cast<size_t>(symbol_size_), total_payload_bytes_ - start);
        if (len > 0) {
            std::memcpy(payload.data() + start, it->second.data(), len);
        }
    }
    return payload;
}

} // namespace Optical::Core
